File: e2e/selenide-selenoid-junit5/src/test/java/step/validation/TasksValidator.java
package step.validation;

import io.qameta.allure.Step;
import org.hamcrest.CoreMatchers;
import page.MainPage;

import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.MatcherAssert.assertThat;

public class TasksValidator {
    MainPage mainPage = new MainPage();

    @Step("Validate task list has task '{title}'")
    public void hasTask(String title) {
        assertThat(mainPage.getTaskTitles(), CoreMatchers.hasItem(title.trim()));
    }

    @Step("Validate task '{title}' is removed from list")
    public void hasNoTask(String title) {
        assertThat(mainPage.getTaskTitles(), not(CoreMatchers.hasItem(title.trim())));
    }
}

File: demo-android/app/src/androidTest/java/com/example/demo/e2e/interaction/validation/TasksValidator.kt
package com.example.demo.e2e.interaction.validation

import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.junit4.AndroidComposeTestRule
import com.example.demo.MainActivity
import com.example.demo.e2e.interaction.page.MainPage
import io.qameta.allure.kotlin.Allure

class TasksValidator(
    rule: AndroidComposeTestRule<*, MainActivity>,
) {
    private val mainPage = MainPage(rule)

    fun hasTask(title: String) {
        Allure.step("Validate task list has task '$title'") {
            val normalizedTitle = title.trim()
            mainPage.waitUntilTaskWithTitlePresent(normalizedTitle)
            mainPage.taskTitleByTitle(normalizedTitle).assertIsDisplayed()
        }
    }

    fun hasNoTask(title: String) {
        Allure.step("Validate task '$title' is removed from list") {
            mainPage.waitUntilTaskWithTitleAbsent(title.trim())
        }
    }
}
